### PR TITLE
Change ParameterResource's "Failed to Start" to "Configuration Missing"

### LINKED ISSUE #3301: Change ParameterResource's "Failed to Start" to "Configuration Missing"
We now have the ability to set both the state text and the icon. We should use a more prescriptive error message and perhaps a log message explaining how to make parameters work.

### PR DESCRIPTION
**Pull request by eerhardt**
Also log the exception message for the Parameter resource so users know what the problem is.

Fix #3301
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3541)

### CODE REVIEW

File: src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -58,13 +61,39 @@ private static string GetParameterValue(IConfiguration configuration, string nam
...
         return builder.AddResource(resource)
                       .WithInitialState(state);
     }
 
+    /// <summary>
+    /// Writes the message to the specified resource's logs.
+    /// </summary>
+    private sealed class WriteParameterLogsHook(ResourceLoggerService loggerService, string resourceName, string message) : IDistributedApplicationLifecycleHook
+    {
+        public Task AfterResourcesCreatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken)
```
Comment (davidfowl) [inline src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:85]: This is too late as it runs after all resources are started.

Comment (eerhardt) [reply src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:85]: Too late for what?

We just want to log the exception message into the log for the resource.

Comment (davidfowl) [reply src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:85]: But you want it to show up quickly, not after all resources have started. If you're provisioning azure resources for e.g. this won't show up until after. These events run after all resources are provisioned.

Comment (davidfowl) [reply src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:85]: BeforeStartAsync works since this resource fails super early.

**Comment by davidfowl** [comment]
Should we also change the source to Configuration?

**Comment by eerhardt** [comment]
> Should we also change the source to Configuration?

I don't know. I kind of like `Parameters:{name}"` showing up in the dashboard... as a indication of what you need to set. Although the log will have this information too.

Screenshot 2024-04-10 at 7 47 18 AM

All the other "Source" strings say where they come from: name of the container, name of the project. It feels consistent to have the name of the parameter here.

**Comment by davidfowl** [comment]
That's what I thought but it seems like nobody figured it out.

File: src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -58,13 +61,39 @@ private static string GetParameterValue(IConfiguration configuration, string nam
...
+        public Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken)
+        {
+            loggerService.GetLogger(resourceName).LogError(message);
+
+            return Task.CompletedTask;
+        }
+
+        public Task AfterEndpointsAllocatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken) => Task.CompletedTask;
+
+        public Task AfterResourcesCreatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken) => Task.CompletedTask;
```
Comment (davidfowl) [inline src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:94]: These aren't needed

**Review by davidfowl — APPROVED** [review]

**Comment by eerhardt** [comment]
> That's what I thought but it seems like nobody figured it out.

The log will really help out.

I don't think making this one inconsistent with the others makes sense. The others have the "name of the source". Unless you mean we preprend "Configuration" to `Parameters:{name}`.

File: src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs (hunk 2 above)
Comment (eerhardt) [reply src/Aspire.Hosting/ParameterResourceBuilderExtensions.cs:94]: image

**Pull request merged**